std/string.cup

```diff
@@ -0,0 +1,268 @@
+
+method slice::get(idx: i32): char{
+    if (__SLICE_LIFETIME_PROTECTION)
+        self::is_str_alive();
+    
+    if (idx < self.len){
+        return self.start[idx];
+    } else {
+        die(here, "slice::get: index out of bounds");
+    }
+}
+
+initialize str(chars: char*){
+    self.chars = malloc(sizeof(char)*strlen(chars)+1);
+    self.len = strlen(chars);
+    strcpy(self.chars, chars);
+}
+
+// Creates a str out of slice
+method slice::to_str(): str*{
+    if (__SLICE_LIFETIME_PROTECTION)
+        self::is_str_alive();
+    
+    let chars: char* = malloc(sizeof(char)*self.len+1);
+    
+    for (let i = 0; i < self.len; ++i){
+        chars[i] = self::get(i);
+    }
+    
+    chars[self.len] = '\0';
+    let str_created = new str(chars);
+    free(chars);
+    return str_created;
+}
+
+method str::push_cstr(chars: char*){
+    self.chars = realloc(self.chars,(sizeof(char)*self.len)+(sizeof(char)*strlen(chars))+1);
+    self.len = self.len + strlen(chars);
+    strcat(self.chars, chars);
+}
+
+method str::push(string: str*){
+    self::push_cstr(string.chars);
+}
+
+method str::push_slice(sl: slice*){
+    self.chars = realloc(self.chars,(sizeof(char)*self.len)+(sizeof(char)*sl.len)+1);
+    let first = self.len;
+    self.len = self.len + sl.len;
+    
+    for (let i = first; i < self.len; ++i){
+        self.chars[i] = sl::get(i-first);
+    }
+    
+    self.chars[self.len] = '\0';
+}
+
+method str::len(): i32{
+    return self.len;
+}
+
+method str::clone(): str*{
+    let clone: str* = new str(self.chars);
+    return clone;
+}
+
+method str::get(idx: i32): char{
+    if (idx < self.len){
+        return self.chars[idx];
+    } else {
+        die(here, "str::get: index out of bounds");
+    }
+}
+
+method str::slice(from: i32, to: i32): slice*{
+    if (to > self.len){
+        die(here, "str::slice: index out of bounds");
+    }
+    if (to < 0 || from < 0){
+        die(here, "str::slice: indexes cannot be negative nums");
+    }
+    
+    // contains index "to" too
+    let s: slice* = new slice(&self.chars[from], (to+1)-from, self); 
+    return s;
+    
+}
+
+method str::split(splitter: char): Vector*{
+    let start_slice = 0;
+    let slices: Vector* = vector_new();
+    for (let i = 0; i < self.len; ++i){
+        if (self::get(i) == splitter){
+            if (!(i == start_slice)){
+                let slice_: slice* = self::slice(start_slice, i-1);
+                slices::push(slice_);
+                start_slice = i + 1;
+            }
+        }
+    }
+    
+    if (start_slice < self.len){
+        let slice_: slice* = self::slice(start_slice, self.len-1);
+        slices::push(slice_);
+    }
+    
+    return slices;
+}
+
+method str::startswith_cstr(prefix: char*): i32{
+    return strstartswith(self.chars, prefix);
+}
+
+method str::startswith(prefix: str*): i32{
+    return self::startswith_cstr(prefix.chars);
+}
+
+method str::equals_cstr(other: char*): i32{
+    return streq(self.chars, other);
+}
+
```

Comment:
This feels weird, Is the `str` type null terminated? I would expect it to not, if it's explicitly storing length.

Reply:
Yeah it has null terminator.

Reply:
Len is length of string without counting the null terminator

Reply:
Essentially you can say that `str` is a wrapper of `char*` and it's `self.chars` are compatible with `cstr` functions.

Reply:
Fair enough

Reply:
It keeps the `len` to not to call `strlen` everytime you need to get string length.